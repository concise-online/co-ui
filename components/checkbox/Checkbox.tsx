import React, { Component, createRef, DOMAttributes, CSSProperties, ChangeEvent, RefObject } from 'react';
import classNames from 'classnames';
import CheckboxGroup from './CheckboxGroup';
import { runCallback } from '../_utils/function';
import { getPrefixCls } from '../_utils/config';

const prefixCls = getPrefixCls('checkbox');

export interface Props extends Omit<DOMAttributes<HTMLInputElement>, 'onChange'> {
  value?: string;
  className?: string;
  checked?: boolean;
  style?: CSSProperties;
  name?: string;
  disabled?: boolean;
  onChange?: (value: string, e: ChangeEvent<HTMLInputElement>) => void;
  forwardedRef?: RefObject<HTMLInputElement>;
}

class Checkbox extends Component<Props> {
  static Group = CheckboxGroup;
  static isCheckbox = true;
  private ref: RefObject<HTMLInputElement>;
  constructor(props: Props) {
    super(props);

    this.ref = createRef();
    this.handleChange = this.handleChange.bind(this);
  }

  focus() {
    (this.ref.current as HTMLInputElement).focus();
  }

  blur() {
    (this.ref.current as HTMLInputElement).blur();
  }

  handleChange(e: ChangeEvent<HTMLInputElement>) {
    const { disabled, onChange } = this.props;

    if (disabled) return;
    runCallback(onChange, e.target.checked, e);
  }

  render() {
    const { children, className, style, disabled, checked, ...checkboxProps } = this.props;
    const classes = classNames(prefixCls, {
      [`${prefixCls}-checked`]: checked,
      [`${prefixCls}-disabled`]: disabled,
    }, className);

    return (
      <label className={classes} style={style}>
        <span className={`${prefixCls}-element`}>
          <input
            {...checkboxProps}
            type="checkbox"
            checked={checked}
            disabled={disabled}
            onChange={this.handleChange}
            ref={this.ref}
          />
        </span>
        <span className={`${prefixCls}-text`}>
          {children}
        </span>
      </label>
    );
  }
}

export default Checkbox;
