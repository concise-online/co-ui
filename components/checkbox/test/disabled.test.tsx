import React from 'react';
import { shallow, render, mount } from 'enzyme';
import Checkbox from '../index';

describe('checkbox disabled feature', () => {
  it('renders correctly', () => {
    const wrapper = render(<Checkbox disabled>Checkbox disabled</Checkbox>);

    expect(wrapper).toMatchSnapshot();
  });

  it('input element has disabled attributes', () => {
    const wrapper = shallow(<Checkbox disabled />);

    expect(wrapper.find('input[disabled]')).not.toBeNull();
  });

  it('set disabled of checkbox', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onChange = jest.fn();
    const wrapper = mount((
      <Checkbox disabled onChange={onChange}>
        Checkbox
      </Checkbox>
    ));

    wrapper.find('input').simulate('change');
    expect(onChange).not.toBeCalled();
  });
});
