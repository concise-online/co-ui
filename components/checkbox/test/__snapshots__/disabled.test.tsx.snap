// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox disabled feature renders correctly 1`] = `
<label
  class="co-checkbox co-checkbox-disabled"
>
  <span
    class="co-checkbox-element"
  >
    <input
      disabled=""
      type="checkbox"
    />
    <span
      class="co-checkbox-inner"
    />
  </span>
  <span
    class="co-checkbox-text"
  >
    Checkbox disabled
  </span>
</label>
`;
